codegen/src/TypeMapper.scala

```diff
@@ -5,6 +5,43 @@ import scala.meta._
 import scala.meta.dialects.Scala33
 import besom.codegen.metaschema._
 
+import scala.annotation.unused
+
+case class TypeToken private (provider: String, module: String, `type`: String, @unused dummy: Boolean = false) {
+  def asString: String = s"${provider}:${module}:${`type`}"
+}
+
+object TypeToken {
+  private val typeTokenFmt: String    = "(.*):(.*)?:(.*)" // provider:module:type
+  private val typeTokenPattern: Regex = ("^" + typeTokenFmt + "$").r
+
+  private def enforceNonEmptyModule(module: String): String =
+    module match {
+      case "" => "index"
+      case _ => module
+    }
+
+  def apply(typeToken: String): TypeToken = typeToken match {
+    case typeTokenPattern(provider, module, typ) =>
+      new TypeToken(
+        provider = provider,
+        module = enforceNonEmptyModule(module),
+        `type` = typ
+      )
+    case _ => throw TypeMapperError(s"Cannot parse type token: $typeToken, typeTokenPattern: $typeTokenPattern")
+  }
+
+  def apply(provider: String, module: String, `type`: String): TypeToken = {
+    new TypeToken(
+      provider = provider,
+      module = enforceNonEmptyModule(module),
+      `type` = `type`
+    )
+  }
+
+  def unapply(t: TypeToken): Option[(String, String, String)] = Some((t.provider, t.module, t.`type`))
```

Comment:
This should be redundant when the public main constructor gets changed to have 3 parameters

Reply:
thank you! :)